pf/internal/convert/defaults.go

```diff
@@ -0,0 +1,23 @@
+package convert
+
+import (
+	"os"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+)
+
+func GetDefaultValue(d *tfbridge.DefaultInfo, resource *tfbridge.PulumiResource) (interface{}, error) {
+	if d == nil {
+		return nil, nil
+	}
+	if d.From != nil {
+		return d.From(resource)
+	} else if d.EnvVars != nil {
+		for _, n := range d.EnvVars {
+			if v := os.Getenv(n); v != "" {
```

Comment:
Should this be os.GetEnv or os.LookupEnv with handling the difference between empty value and no value in the env?

Reply:
Yes, you're right. But I was lazy and copied the code around the EnvVars from the TF Bridge implementation about handling default values. I'll change this 👍🏼

Reply:
Eh, the incumbent code uses so perhaps that's what we should keep. "" is missing.

```
func multiEnvDefault(names []string, dv interface{}) interface{} {
	for _, n := range names {
		if v := os.Getenv(n); v != "" {
			return v
		}
	}
	return dv
}

```

Reply:
Incumbent code does schema-aware parsing for TypeBool, TypeInt, TypeFloat:

https://github.com/pulumi/pulumi-terraform-bridge/blob/9baf57a1b25518606d53a509b10dc492d99ce229/pkg/tfbridge/schema.go#L620-#L620

Reply:
Is this really necessary? Because in the newly introduced `SetDefaultValues` I'll use the value from the environment variable (if present) and call `resource.NewPropertyValue` to convert the input value to a proper property value, as you proposed.

Reply:
@t0yv0 Yep, you're right. I thought `resource.NewPropertyValue` will handle this and convert a `string` to a the target type if they differ. But it does not. That's why the code at https://github.com/pulumi/pulumi-terraform-bridge/blob/9baf57a1b25518606d53a509b10dc492d99ce229/pkg/tfbridge/schema.go#L620-#L620 is doing this.

Issue is that I need the target type to do the conversation and this seems to be very different between PF and the V2 or V1 implementations.

If I understand the code in PF correctly I do the following:

(1) get a `resourceHandle` to the current resource
(2) use `rh.Attr[propertyName].GetType()` to get the type
(3) use cascaded `if` statements to check for the target type

`if tftypes.Is(tftypes.String) {
  ...
} else if tftypes.Is(tftypes.Number) {
} else if tftypes.Is(tftypes.Bool) {
}

![image](https://user-images.githubusercontent.com/14177833/235469974-4f28bd46-1dd2-4b68-a8e2-89c3dc93c344.png)

Is this correct?

Reply:
Ahhh ... just spotted `AttributeAtPath(context.Context, path.Path) (Attr, diag.Diagnostics)` 😁 
Issue still: what is the path of `sql_auth -> username`? And how can it be "computed" when we walk over the fields recursively?

Reply:
Came up with something like this

```go
func SetDefaultValues(res *tfbridge.PulumiResource, schema pfutils.Schema, fields map[string]*tfbridge.SchemaInfo) error {
  for key, fld := range fields {
    if _, ok := res.Properties[resource.PropertyKey(key)]; !ok && fld.Default != nil {
      // using default value for empty property
      v, err := getDefaultValue(fld.Default, res)
      if err != nil {
        return err
      }
      if pval, ok := v.(resource.PropertyValue); ok {
        res.Properties[resource.PropertyKey(key)] = pval
      } else {
        attr, ok := schema.Attrs()[key]
        if !ok {
          return fmt.Errorf("failed to get attribute from schema for key %s", key)
        }
        switch attr.GetType().(type) {
        case basetypes.Int64Type:

        case basetypes.Float64Type:

        case basetypes.BoolType:

        case basetypes.StringType:

        default:
          return fmt.Errorf("unable to convert value of type %T to attribute of type %T", v, attr)
        }
        res.Properties[resource.PropertyKey(key)] = resource.NewPropertyValue(v)
      }
    }
  }

  return nil
}
```

Reply:
I think it would be very annoying to work with AttributeAtPath in this code because Pulumi paths do not map to TF paths 1-1, there's cases of nested renaming and maxitems=1 flattening to be handled. 

I'm working on a WIP PR to find some easy to use API to hide this complexity so we don't need to deal with it here, but it's not quite done yet. Hoping to finish tomorrow.

https://github.com/pulumi/pulumi-terraform-bridge/pull/1065/commits/3a89a67168bc37816f76ccc2a3c75af8eae50dc4#diff-0d24b3420725c5b1914ff5deafbaeee99714952daa359bdc2f8846aec47463f6R34

Reply:
Regarding env-variable parsing, rh.Attr[propertyName].GetType() and inferring Pulumi type from the PF type works for simple type like int,bool,float, so that should work! However it would be also easy for me to add some plumbing so you get the shim.Type model for these types to work with also.

Reply:
I realize more and more that I always approach the subject too simply in my mind. If you look at environment variables, then you can assume that only simple values are defined. Or do we also want to parse JSON from a string or something similar? I think that can't be the objective.  But do we need to convert a default value that was simply defined in code?

Reply:
I think for this PR to go in we only need to match the sophistication of the incumbent v3 bridge which doesn't do objects (JSON parsing) as yet. We can backlog the rest as a feature enhancement.